packages/react-native/ReactCommon/react/performance/timeline/CircularBuffer.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+#include "PerformanceEntry.h"
+
+namespace facebook::react {
+
+constexpr size_t DEFAULT_MAX_SIZE = 1024;
+
+/**
+ * A container for storing entries of type T, with the following properties:
+ * - It can only grow up to a specified max size
+ * - It's a circular buffer (the oldest elements are dropped if reached max
+ * size and adding a new element)
+ *
+ * Note that the space for maxSize elements is reserved on construction. This
+ * ensures that pointers to elements remain stable across add() operations.
+ */
+template <class T>
+class CircularBuffer {
+ public:
+  explicit CircularBuffer(size_t maxSize = DEFAULT_MAX_SIZE)
+      : maxSize_(maxSize) {
+    entries_.reserve(maxSize_);
+  }
+
+  /**
+   * Adds (pushes) element into the buffer.
+   *
+   * Returns the result of the operation, which will depend on whether the
+   * buffer reached the max allowed size, in which case `true` is returned. If
+   * no items were overridden `false` is returned.
+   */
+  bool add(const T&& el) {
+    if (entries_.size() < maxSize_) {
+      // Haven't reached max buffer size yet, just add and grow the buffer
+      entries_.emplace_back(el);
+      return false;
+    } else {
+      // Overwrite the oldest (but already consumed) element in the buffer
+      entries_[position_] = el;
+      position_ = (position_ + 1) % entries_.size();
+      return true;
+    }
+  }
+
+  /**
+   * Returns pointer to next entry which would be overwritten or dropped if
+   * added a new element. Null if no entry will be dropped.
+   */
+  const T* getNextOverwriteCandidate() const {
```

Comment:
This doesn't seem like it should be part of the API. For testing, we can just check the contents of the buffer to see which one would be overwritten.